addon-test-support/-private/ember-exam-qunit-test-loader.js

```diff
@@ -0,0 +1,142 @@
+   * Loads the test modules depending on the urlParam
+   */
+  loadModules() {
+    const loadBalance = this._urlParams.loadBalance;
+    let partitions = this._urlParams.partition;
+    let split = parseInt(this._urlParams.split, 10);
+
+    split = isNaN(split) ? 1 : split;
+
+    if (partitions === undefined) {
+      partitions = [1];
+    } else if (!Array.isArray(partitions)) {
+      partitions = [partitions];
+    }
+
+    super.loadModules();
+
+    if (loadBalance) {
+      this.setupLoadBalanceHandlers();
+      this._testModules = splitTestModules(
+        weightTestModules(this._testModules),
+        split,
+        partitions
+        );
+      this._testem.emit('testem:set-modules-queue', this._testModules);
+    } else {
+      this._testModules = splitTestModules(this._testModules, split, partitions);
+      this._testModules.forEach((moduleName) => {
+        super.require(moduleName);
+        super.unsee(moduleName);
+      });
+    }
+  }
+
+  /**
+   * Allow loading one module at a time.
+   *
+   * @param {string} moduleName
+   */
+  loadIndividualModule(moduleName) {
+    if (moduleName === undefined) {
+      throw new Error(
+        'Failed to load a test module. `moduleName` is undefined in `loadIndividualModule`.'
+      );
+    }
+    super.require(moduleName);
+    super.unsee(moduleName);
+  }
+
+  /**
+   * setupLoadBalanceHandlers() registers QUnit callbacks neeeded for the load-balance option.
+   */
+  setupLoadBalanceHandlers() {
+    // this handler sets up testem event handlers to load a test module. The cleanup of handlers
+    // ensures each event will only envoke one handler.
+    // if no cleanup was done, multiple handlers will be envoked for a single event, causing
+    // multiple get-next-module events, leading to uneven load balancing
+    const nextModuleHandler = (resolve , reject) => {
+      const getTestModule = (moduleName) => {
+        try {
+          this.loadIndividualModule(moduleName);
+
+          // if no tests were added, request the next module
+          if (this._qunit.config.queue.length === 0) {
+            this._testem.emit('testem:get-next-module');
+          } else {
+            // `removeEventCallbacks` removes if the event queue contains the same callback for
+            // an event.
+            this._testem.removeEventCallbacks('testem:next-module-response', getTestModule);
+            this._testem.removeEventCallbacks('testem:module-queue-complete', moduleComplete);
+            resolve();
+          }
+        } catch (err) {
+          reject(err);
+        }
+      }
+      const moduleComplete = () => {
+        this._testem.removeEventCallbacks('testem:next-module-response', getTestModule);
+        resolve();
+      }
+
+      this._testem.on('testem:next-module-response', getTestModule);
+      this._testem.on('testem:module-queue-complete', moduleComplete);
+      this._testem.emit('testem:get-next-module');
+    }
+
+    this._qunit.begin(() => {
```

Comment:
@choheekim / @step2yeung when I remove this `this._qunit.begin(...)` the tests continue to pass, so either it is redundant or the tests are missing the scenario a where it is needed.

Reply:
Working on investigating to clarify use cases of having to have `this._qunit_being(...)`

Reply:
thanks

Reply:
in ember-qunit `Ember.onerror` test is being added to QUnit module and executed. As the test is complete `moduleDone` callbacks is called and that's why we see tests continue to pass. (fyi, https://github.com/emberjs/ember-qunit/blob/faddd902d313721c33f20dd80a225946fbbd7979/addon-test-support/ember-qunit/index.js#L145) However, with `setupEmberOnerrorValidation: false` option to disable `Ember.onerror` test `moduleDone` callback is called and it results a test failing with `Browser Id 1 - global failure
    ---
        actual: >
            [object Object]
        stack: >
            Error: No tests were run.`

To prevent the incident, we have `this._qunit.begin(..)` to emit an event to emit to get a first test module to run.

Reply:
Is there a way we can ensure there is test coverage for this behavior (or an inline comment)? The fact that it can be removed and tests pass is very risky.

Reply:
I will add an inline comment to explain why `this._qunit.begin()` callback has to be registered and think of if we can write some tests against it